### PR TITLE
feat: distributed copy into table  from query

### PR DESCRIPTION
**Pull request by JackTan25**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary
distain
Summary about this PR
finish distributed copy into table from query
- Closes #issue
https://github.com/datafuselabs/databend/issues/11752 https://github.com/datafuselabs/databend/issues/8594 https://github.com/datafuselabs/databend/issues/6395

### CODE REVIEW

**Comment by JackTan25** [comment]
askbend:summary

**Comment by JackTan25** [comment]
askbend:summary

**Comment by BohuTANG** [comment]
askbend:summary

**Comment by JackTan25** [comment]
askbend:summary

File: src/query/sql/src/executor/physical_plan_display.rs [hunk 1]
```diff
@@ -82,9 +83,12 @@ impl<'a> Display for PhysicalPlanIndentFormatDisplay<'a> {
             PhysicalPlan::ProjectSet(unnest) => write!(f, "{}", unnest)?,
             PhysicalPlan::RuntimeFilterSource(plan) => write!(f, "{}", plan)?,
             PhysicalPlan::RangeJoin(plan) => write!(f, "{}", plan)?,
-            PhysicalPlan::DistributedCopyIntoTable(copy_into_table) => {
+            PhysicalPlan::DistributedCopyIntoTableFromStage(copy_into_table) => {
                 write!(f, "{}", copy_into_table)?
```
Comment (RinChanNOWWW) [inline src/query/sql/src/executor/physical_plan_display.rs:87]: copy_into_table_from_stage?

Comment (JackTan25) [reply src/query/sql/src/executor/physical_plan_display.rs:87]: yes, we have two cases, copy into table from stage like copy into t1 from @s1, secondly, copy into t from (select ....)

Comment (JackTan25) [reply src/query/sql/src/executor/physical_plan_display.rs:87]: > yes, we have two cases, copy into table from stage like copy into t1 from @s1, secondly, copy into t from (select ....)
we need to uniform these as one plan in the future. but they are not compatible now, from stage will be changed to from query by @youngsofun later.

Comment (RinChanNOWWW) [reply src/query/sql/src/executor/physical_plan_display.rs:87]: So if the display content should be "copy_into_table_from_stage".

Comment (JackTan25) [reply src/query/sql/src/executor/physical_plan_display.rs:87]: sure!!! thanks

**Review by youngsofun — APPROVED** [review]

**Review by RinChanNOWWW — APPROVED** [review]

File: src/query/service/src/interpreters/interpreter_copy.rs [hunk 2]
```diff
@@ -302,7 +306,67 @@ impl CopyInterpreter {
             .await?;
 
         let (mut build_res, source_schema, files) = if let Some(query) = &plan.query {
-            let (build_res, source_schema) = self.build_query(query).await?;
+            let mut build_res;
+            let (select_interpreter, source_schema) = self.build_query(query).await?;
+            let plan_query = select_interpreter.build_physical_plan().await?;
```
Comment (BohuTANG) [inline src/query/service/src/interpreters/interpreter_copy.rs:311]: If COPY is distributed, how we gather the status from the cluster through the coornidate query node?

File: src/query/sql/src/planner/optimizer/optimizer.rs [hunk 3]
```diff
@@ -135,21 +135,23 @@ pub fn optimize(
...
-                            opt_ctx.config.enable_distributed_optimization
-                                && ctx.get_settings().get_enable_distributed_copy()?;
-                        info!(
-                            "after optimization enable_distributed_copy? : {}",
-                            into_table.enable_distributed
-                        );
-                        CopyPlan::IntoTable(into_table)
+
+                CopyPlan::IntoTable(mut into_table) => {
+                    into_table.enable_distributed = opt_ctx.config.enable_distributed_optimization
```
Comment (JackTan25) [inline src/query/sql/src/planner/optimizer/optimizer.rs:140]: we get cluster nodes info here and to enable distributed copy. @BohuTANG

File: src/query/service/src/interpreters/interpreter_copy.rs (hunk 2 above)
Comment (JackTan25) [reply src/query/service/src/interpreters/interpreter_copy.rs:311]: here is just a physical plan, we will use exchange physical plan at line 342

File: src/query/service/src/pipelines/builders/copy.rs [hunk 4]
```diff
@@ -48,7 +49,10 @@ use crate::sessions::QueryContext;
 
 pub enum CopyPlanType {
     CopyIntoTablePlanOption(CopyIntoTablePlan),
-    DistributedCopyIntoTable(DistributedCopyIntoTable),
+    DistributedCopyIntoTableFromStage(DistributedCopyIntoTableFromStage),
+    // also distributed plan, but we think the real disteibuted part is the query
```
Comment (zhyass) [inline src/query/service/src/pipelines/builders/copy.rs:53]:
```suggestion
    // also distributed plan, but we think the real distributed part is the query
```

File: src/query/service/src/interpreters/interpreter_copy.rs [hunk 5]
```diff
@@ -366,21 +398,78 @@ impl CopyInterpreter {
...
-        let build_res = build_distributed_pipeline(&self.ctx, &exchange_plan, false).await?;
+        let catalog_name;
+        let database_name;
+        let table_name;
+        let stage_info;
+        let files;
+        let force;
+        let purge;
+        let is_overwrite;
+        let mut build_res = match distributed_plan {
```
Comment (RinChanNOWWW) [inline src/query/service/src/interpreters/interpreter_copy.rs:411]: I think you can merge these `let`s into one `let` to make codes prettier.

File: src/query/service/src/pipelines/builders/copy.rs (hunk 4 above)
Comment (JackTan25) [reply src/query/service/src/pipelines/builders/copy.rs:53]: thanks

File: src/query/service/src/interpreters/interpreter_copy.rs (hunk 5 above)
Comment (JackTan25) [reply src/query/service/src/interpreters/interpreter_copy.rs:411]: thanks

**Review by youngsofun — APPROVED** [review]

File: src/query/sql/src/planner/optimizer/optimizer.rs [hunk 6]
```diff
@@ -135,21 +135,21 @@ pub fn optimize(
...
+                        && ctx.get_settings().get_enable_distributed_copy()?;
+                    match into_table.query {
+                        Some(_) => CopyPlan::IntoTable(into_table),
+                        None => {
+                            info!(
+                                "after optimization enable_distributed_copy? : {}",
+                                into_table.enable_distributed
+                            );
+                            CopyPlan::IntoTable(into_table)
+                        }
```
Comment (dantengsky) [inline src/query/sql/src/planner/optimizer/optimizer.rs:150]:
```suggestion
                        && ctx.get_settings().get_enable_distributed_copy()?;
                    info!(
                                "after optimization enable_distributed_copy? : {}",
                                into_table.enable_distributed
                    );
                    CopyPlan::IntoTable(into_table)
```

**Review by dantengsky — APPROVED** [review]
LGTM.

suggested a tiny code refactoring, hope I get it correct.

**Pull request merged**